pkgs/unified_analytics/USAGE_GUIDE.md

```diff
@@ -122,11 +136,43 @@ final Analytics analytics = Analytics(...);
 print('This user's status: ${analytics.telemetryEnabled}');  // true if opted-in
 ```
 
+## Checking for New Versions of Consent Message
+
+In the event that the package consent messaging needs has been updated, an
+API has been exposed on an instance of `Analytics` that will notify the tool
+using this package whether to display the message again.
+
+```dart
+// Begin by initializing the class
+//
+// This is assuming that the tool has already been onboarded
+// and that the user has already seen the previous version of
+// the consent message
+final Analytics analytics = Analytics(...);
+
+
+// Check the instance's method on if it should show
+// a message
+if (analytics.shouldShowMessage) {
```

Comment:
This doesn't work. We can't create an instance of `Analytics` in order to find out whether we need to show the message if creating an instance will assume that the message will have been seen by the user. For the analysis server case we need to wait until the user has clicked "OK" before we can set the state from "needs to show message" to "can collect analytics on the next run".

I think you need to change the code so that creating an instance doesn't change the state of the system, and add an explicit "user has seen the message" method whose purpose is to change the state.

If you do that, then there's no need for the top-level functions because everything can be done through an instance of `Analytics`. That will also reduce the probability that the state will change (via a different process) and not be noticed.

Reply:
Understood, but this is not for that case here. This section is when the version of the consent message has changed for whatever reason we may have. We need to run the new message to them – however, in this instance, we would have already had an instance created.

There is another section in the usage guide above this one where we now have a static method to retrieve the message for a given tool without having to create an instance of it. It is labeled `Displaying Consent Message to Users`

Reply:
@eliasyishak Creating an Analytics instance still should not have a side effect of the message being assumed to be sent.  Is there a reason why this shouldn't be reworked so that you can create an Analytics instance, ask it whether or not a message needs to be sent, then tell the Analytics instance that you sent the message?  Only then should state be updated on disk, and we can do away with static methods.

Reply:
> Creating an Analytics instance still should not have a side effect of the message being assumed to be sent.

I'd state that more strongly: it can't have that side effect.

In the IDE case, the analysis server can determine whether it's possible to display a message before doing anything else, but if a message can be shown it will be required to create an instance of `Analytics` in order to determine whether a message needs to be displayed. It can then send the message to the IDE, but we can only count it as displayed when the user clicks the 'OK' button.

Reply:
That's a very valid point, I'll rework this so that we can remove the static methods and use the two step approach you laid you mentioned @jcollins-g – thank you both for your input!